sidechain/block-composer/src/block_composer.rs

```diff
@@ -37,133 +36,106 @@ use std::{format, marker::PhantomData, sync::Arc, vec::Vec};
 
 /// Compose a sidechain block and corresponding confirmation extrinsic for the parentchain
 ///
-pub trait ComposeBlockAndConfirmation {
+pub trait ComposeBlockAndConfirmation<Externalities, PB: BlockT> {
 	type SidechainBlockT: SignedBlockT;
-	type ParentchainBlockT: BlockT;
```

Comment:
Removed ParentchainBlockT from types but made it generic instead. 

Reasoning: ComposeBlock can be very well used for different parentchain block types, doesn't need to be explicitly one type.

Reply:
Well one is just syntactic sugar for the other, so nothing really changed? They're both ways to make this trait generic. One is through a type parameter, the other through an associated type - but are generic and interchangeable.

Reply:
The Rust guidelines mention associated types especially in combination with output types, which here is only `SidechainBlockT`. So from that perspective your change is fine - although it's just a minor style change ☺️

Reply:
No, they are not. I had to fight with this very problem: Imagine a two functions with an associated State type each. 
FunctionOne does something on the state and return it. The returned state type is now FunctionOne::State.
FunctionTwodoes something on the state and return it. The returned state type is now FunctionTwo::State.

And now imagine a third function that should do something with each state and also has an assosciated type  - it won't accept the two different State types, that have been returned by the functions above.

A bit more general:

According to: https://blog.thomasheartman.com/posts/on-generics-and-associated-types

> **Generics** allow you to implement the same trait numerous times for the same type by changing the type parameter. The From<T> trait was mentioned earlier, and it's a great example of this. Because From<T> uses a generic type, we can implement it for any number of type parameters.
> 
> For instance, say you have a type MyNumeric. You can implement From<u8>, From<u16>, From<u32>, and so on. This makes generics very useful if it makes sense to have multiple trait implementations varying only in type parameters.
> 
> **Associated types,** on the other hand, only allow a single implementation. Because a type can only implement a trait once, this can be used to constrain the number of implementations.
> 
> The Deref trait comes to mind. Deref has an associated type Target, which it can be dereferenced to. It would get really confusing if a type could implement Deref into an arbitrary set of other types (and probably very tricky for type inference!).
> 
> Because a trait can only be implemented once per type, associated types also offer some notational benefits. Using associated types means you don't have to add type annotations for all the extra types. This is touted as an engineering benefit in the RFC.

Reply:
Okay, then I was wrong - wasn't aware of this distinction. Although I don't think I understand the example you give, I'd have to see it in code 😄

Reply:
Ah yeah, sorry. I was a little desperate because of it, so I might have left earth a little there 😋 
But I think this sentence is the most important one: _Associated types, on the other hand, only allow a single implementation._

So if we want to use to enforce this - like for sidechain block I think that makes perfectly sense to have only one sidechain block type - if not, simple generics should be preferred. At least that's my take from it.

Reply:
Okay then the question is: Do we have different parentchain block types now already? AFAIK we have one type defined in `itp_types`?

Reply:
Usually in our implementations, like here, these types are then defined by generic type parameters - so they allow multiple implementations, even if the trait has an associated type. So I guess I still don't understand 😛 But the change is fine for me.

Reply:
Okay, then lets go with the style guide for now? Output as types, Input as generics in traits?

Reply:
Remark: _Associated types, on the other hand, only allow a single implementation from the SAME struct/type (Self)_. 

Example where assosciated types would not work: `impl From<Error> for sgx_status_t`. https://github.com/integritee-network/worker/blob/master/enclave-runtime/src/error.rs#L43
We want to have multiple implementations of From for the `sgx_status_t`